qadence/backends/adjoint.py

```diff
@@ -0,0 +1,117 @@
+                ):
+                    mu = apply_operator(
+                        ctx.out_state, op.jacobian_generator(values), op.qubit_support
+                    )
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+                elif time_param.requires_grad:
+                    mu = apply_operator(ctx.out_state, op.jacobian_time(values), op.qubit_support)
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+                ctx.projected_state = apply_operator(
+                    ctx.projected_state, op.dagger(values), op.qubit_support
+                )
+            elif isinstance(op, ScalePyQOperation):
+                ctx.out_state = apply_operator(ctx.out_state, op.dagger(values), op.qubit_support)
+                scaled_pyq_op = op.operations[0]
+                if (
+                    isinstance(scaled_pyq_op, PyQParametric)
+                    and values[scaled_pyq_op.param_name].requires_grad
+                ):
+                    mu = apply_operator(
+                        ctx.out_state,
+                        scaled_pyq_op.jacobian(values),
+                        scaled_pyq_op.qubit_support,
+                    )
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+
+                if values[op.param_name].requires_grad:
+                    grads.append(2 * -values[op.param_name])
+                ctx.projected_state = apply_operator(
+                    ctx.projected_state, op.dagger(values), op.qubit_support
+                )
+            elif isinstance(op, PyQCircuit):
+                grads = [g for sub_op in op.reverse() for g in _apply_adjoint(ctx, sub_op)]
+            elif isinstance(op, (PyQPrimitive)):
+                ctx.out_state = apply_operator(ctx.out_state, op.dagger(values), op.qubit_support)
+                if isinstance(op, (PyQParametric)) and values[op.param_name].requires_grad:
+                    mu = apply_operator(
+                        ctx.out_state,
+                        op.jacobian(values),
+                        op.qubit_support,
+                    )
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+                ctx.projected_state = apply_operator(
+                    ctx.projected_state, op.dagger(values), op.qubit_support
+                )
+            else:
+                raise TypeError(
+                    f"AdjointExpectation does not support a backward pass for type {type(op)}."
+                )
+
+            return grads
+
+        grads = list(
+            reversed(
+                [grad_out * g for op in ctx.circuit.reverse() for g in _apply_adjoint(ctx, op)]
+            )
+        )
+        num_grads = len(grads)
+        num_params = len(ctx.saved_tensors)
+        diff = num_params - num_grads
+        grads = grads + [tensor([0]) for _ in range(diff)]
+        # Set observable grads to 0
```

Comment:
If you would like to have reverse mode for the observable parameters, I guess that the parameters should not even enter into this function, right?

Reply:
for now they have to since we store all parameter values in the same dict (both circuit and observable). we can split them in the future and then have separate passes for each

Reply:
https://github.com/pasqal-io/qadence/issues/188